service/dap/server.go

```diff
@@ -1136,25 +1137,34 @@ func (s *Session) onDisconnectRequest(request *dap.DisconnectRequest) {
 		// This is a multi-use server/debugger, so a disconnect request that doesn't
 		// terminate the debuggee should clean up only the client connection and pointer to debugger,
 		// but not the entire server.
-		status := "halted"
-		if s.isRunningCmd() {
+		status := "running"
+		if request.Arguments.SuspendDebuggee {
+			status = "suspended"
+			s.changeStateMu.Lock()
+			defer s.changeStateMu.Unlock()
```

Comment:
This defer gets executed when the function exits. Is this really the intended scope? There's a lot of code after this if/else that will get executed either with or without holding the mutex.